### PR TITLE
Class modifiers for apis

### PR DESCRIPTION
**Pull request by munificent**
I'm sure this will need a decent amount of work to beat it into shape, but I think it's ready for review.

Is there something I need to change to get this to show up under "Resources" on the side navigation?

### CODE REVIEW

File: src/resources/class-modifiers-for-apis.md
Comment (eernstg) [inline src/resources/class-modifiers-for-apis.md:208]: Perhaps call it a 'constraint' rather than a 'limitation': It's actually fair to say that it is a limitation in those other languages, but in the context of Dart it is a deliberate choice: "I want the complete package consisting of some guarantees and some constraints", and if the constraint/limitation part is a problem then just don't write `base`.

**Comment by eernstg** [comment]
@munificent, this is great, shape or non-shape! ;-)

**Comment by MaryaBelanger** [comment]
> Is there something I need to change to get this to show up under "Resources" on the side navigation?

Yup, just add a `title` and `permalink` bullet for it here: https://github.com/dart-lang/site-www/blob/main/src/_data/side-nav.yml#L324

Taking a look at the rest of it now, thanks for doing this!

**Comment by MaryaBelanger** [comment]
@munificent could you add a note to the main class modifiers page (`/language/class-modifiers.md`) as part of this PR? Wording variable, something like:

```
{{site.alert.note}}
  If you maintain a library, read the
  [Class modifiers for API maintainers][/resources/class-modifiers-for-apis]
  page for guidance on how to navigate these changes for your libraries. 
{{site.alert.end}}
```

I tried it at line 31 (the end of the intro section) just because there's already a version note at the top of the page, so it _kind of_ "looks better". You can do that or add it wherever you want, top of the page alongside the version note is fine too (and of course visibility is important so no point in hiding it just for looks)

**Review by MaryaBelanger — COMMENTED** [review]
I didn't get all the way through yet, but nothing major anyway. 

Taking a break to add a couple thoughts:
- Would it make sense or be a nice plus to have each of the modifier sections in the regular class modifiers page link to their corresponding section on this page? I know the audiences are different, but I think just having "To learn more about (modifier) for library maintainers, check out..."
- I wonder if this would belong more under the Development or Packages sidenav section. Granted I couldn't really summarize the purposes of those sections, it just seems like this info might be useful to people who are concerned about package stuff or development stuff? Your call

File: src/resources/class-modifiers-for-apis.md
Comment (MaryaBelanger) [inline src/resources/class-modifiers-for-apis.md:194]: "also" or "too", not both
```suggestion
must also prevent being implemented.
```

File: src/resources/class-modifiers-for-apis.md
Comment (MaryaBelanger) [inline src/resources/class-modifiers-for-apis.md:199]: Shorten sentence
```suggestion
Applying `base` to a type requires some care, then.
It affects not just what users can do with your class or mixin,
```

File: src/resources/class-modifiers-for-apis.md
Comment (MaryaBelanger) [inline src/resources/class-modifiers-for-apis.md:262]:
```suggestion
    This way, the compiler can find them all. It knows that there aren't
```

File: src/resources/class-modifiers-for-apis.md
Comment (MaryaBelanger) [inline src/resources/class-modifiers-for-apis.md:290]: I would make this it's own section (it'd be nice to point people to directly). Either top level or a subsection of sealed:

```suggestion

## `sealed` versus `final`

```

**Review by MaryaBelanger — APPROVED** [review]
Approved % the minor comments I made. 

I would really like this to be a blog post too, maybe it would reach more people that way? It kind of missed the Dart 3 announcement so I worry about it's visibility. It's great content and should be front and center but the site isn't really conducive to that. 

I don't want to add more work, but I would be ok with it:
- being both a page on the site AND a blog post, to balance visibility and longevity
  - that's probably bad practice but I feel like this is a special case...
- being just a blog post OR just a page on the site, if the work to achieve either is equivalent
  - we're obviously further along with adding it to the site, but if you want to choose to make it a blog post, I will personally steam roll it ahead and make sure it's published ASAP, just let me know. 

Edit: a review from @lrhn would be great before this gets published, or after if timing is a concern.

File: src/resources/class-modifiers-for-apis.md
Comment (MaryaBelanger) [inline src/resources/class-modifiers-for-apis.md:85]:
```suggestion
- If the answer to both is "yes", then make it a mixin class. 
- If the answer to the second is "no", then just leave it as a class. 
- If the answer to the first is "no" and the second is "yes",
  then change it from a class to a mixin declaration.
```

**Comment by munificent** [comment]
Thank you so much for picking this after I got sucked into the formatter and dropped it onto the floor.

I like the idea of it being both a blog post and a permanent guide. Maybe we could make the blog post be a condensed version that takes you to the full guide if you want to know more? @mit-mit might have an opinion.

**Comment by MaryaBelanger** [comment]
I'm going to count my review as the copy review and merge this in. I'll follow up with @mit-mit about a possible blog post. Thanks!

**Pull request merged**
